### ISSUE TITLE
hello.py worked ... until it broke in Ch. 7 

### ISSUE THREAD

**Issue by alwayslearnedstuff** [miguelgrinberg/flasky#320]
I'm typing code from the book and the app worked fine in Cloud9 IDE until chapter 7 when I split hello.py into the larger app structure; then the app broke. 

When I go to run `python manage.py shell` from _bash_, I'm running into this error:
```
  File "manage.py", line 9, in <module>
     app = create_app(os.getenv('FLASK_CONFIG') or 'default')
  File "/home/ubuntu/workspace/app/__init__.py", line 32, in create_app
     from .main import main as main_blueprint
ImportError: No module named main
```

Cloud9 IDE requires setting port and host variables to run hello.py (code below). 
Could the "manage.py" line 9 error above be a result of not setting those port & host variables?  
```
if __name__ == "__main__":
  port = int(os.getenv('PORT', 8080))
  host = os.getenv('IP', '0.0.0.0')
  app.run(port=port, host=host)
  app.debug = True

```

**Comment by miguelgrinberg** [comment]
The error claims that you don't have a module named `main` in your project. It's not related to how your start the project.

**Comment by alwayslearnedstuff** [comment]
Thanks for the reply and I originally thought so too. I spent about an hour debugging before posting here and, e.g. I found the book's errata and fixed that line of code from error to be `from main...` to `from .main  ... `   
However... I still get that error indicating no main module, and the fact is my file structure is exactly like you outlined for the large app structure: main is a folder under app. Here's code of the main 3 files: 

app/main/__init__.py
```
# subpackage in app
from flask import Blueprint

main = Blueprint('main', __name__)

from . import views, errors
```
config.py
```
import os, time
basedir = os.path.abspath(os.path.dirname(__file__))

class Config:
  SECRET_KEY = os.environ.get('SECRET_KEY') or 'hard to guess string'
  SQLALCHEMY_COMMIT_ON_TEARDOWN = True
  FLASKY_MAIL_SUBJECT_PREFIX = '[Flasky]'
  FLASKY_MAIL_SENDER = 'Flasky Admin <flasky@example.com>'
  FLASKY_ADMIN = os.environ.get('FLASKY_ADMIN')

  @staticmethod
  def init_app(app):
    pass
  
class DevelopmentConfig(Config):
  DEBUG = True
  MAIL_SERVER = 'smtp.googlemail.com'
  MAIL_PORT = 587
  MAIL_USE_TLS = True
  MAIL_USERNAME = os.environ.get('MAIL_USERNAME')
  MAIL_PASSWORD = os.environ.get('MAIL_PASSWORD')
  SQLALCHEMY_DATABASE_URI = os.environ.get('DEV_DATABASE_URL') or \
  'sqlite:///' + os.path.join(basedir, 'data-dev.sqlite')
  
class TestingConfig(Config):
  TESTING = True
  SQLALCHEMY_DATABASE_URI = os.environ.get('TEST_DATABASE_URL') or \
  'sqlite:///' + os.path.join(basedir, 'data-test.sqlite')

class ProductionConfig(Config):
  SQLALCHEMY_DATABASE_URI = os.environ.get('DATABASE_URL') or \
  'sqlite:///' + os.path.join(basedir, 'data.sqlite')

config = {
  'development': DevelopmentConfig,
  'testing': TestingConfig,
  'production': ProductionConfig,
  'default': DevelopmentConfig
}
```

manage.py
```
#!/usr/bin/env python
import os
from app import create_app, db
from app.models import User, Role
from flask.ext.script import Manager, Shell
from flask.ext.migrate import Migrate, MigrateCommand
from flask.ext.sqlalchemy import SQLAlchemy

port = int(os.getenv('PORT', 8080))
host = os.getenv('IP', '0.0.0.0')
app = create_app(os.getenv('FLASK_CONFIG') or 'default')
manager = Manager(app)
migrate = Migrate(app, db)

#@app.shell_context_processor
def make_shell_context():
  return dict(app=app, db=db, User=User, Role=Role)
manager.add_command("shell", Shell(make_context=make_shell_context))
manager.add_command('db', MigrateCommand)

@manager.command
def test():
  """Run the unit tests."""
  import unittest
  tests = unittest.TestLoader().discover('tests')
  unittest.TextTestRunner(verbosity=2).run(tests)

if __name__ == "__main__":
  manager.run(port=port, host=host)
#	app.run()
# port = int(os.getenv('PORT', 8080))
# host = os.getenv('IP', '0.0.0.0')
# app.run(port=port, host=host)
# app.debug = True
```

**Comment by miguelgrinberg** [comment]
Where do you get the error about the main module? Can I see the complete stack trace?

**Comment by alwayslearnedstuff** [comment]
1. Sure.  It's basically what I posted earlier b/c rest of stack trace just has messages regarding outdated modules that don't affect program execution (i.e. these outputted prior to ch7 when hello.py worked).  
But here's the complete stack trace when I have `from . import db` commented out: 
```
(venv)alwayslearnedstuff:~/workspace $ python manage.py shell                                                                                                                
/home/ubuntu/workspace/app/__init__.py:3: ExtDeprecationWarning: Importing flask.ext.script is deprecated, use flask_script instead.
  from flask.ext.script import Manager
/home/ubuntu/workspace/app/__init__.py:5: ExtDeprecationWarning: Importing flask.ext.bootstrap is deprecated, use flask_bootstrap instead.
  from flask.ext.bootstrap import Bootstrap
/home/ubuntu/workspace/app/__init__.py:6: ExtDeprecationWarning: Importing flask.ext.moment is deprecated, use flask_moment instead.
  from flask.ext.moment import Moment
/home/ubuntu/workspace/app/__init__.py:7: ExtDeprecationWarning: Importing flask.ext.wtf is deprecated, use flask_wtf instead.
  from flask.ext.wtf import Form
/home/ubuntu/workspace/app/__init__.py:8: ExtDeprecationWarning: Importing flask.ext.sqlalchemy is deprecated, use flask_sqlalchemy instead.
  from flask.ext.sqlalchemy import SQLAlchemy
/home/ubuntu/workspace/app/__init__.py:9: ExtDeprecationWarning: Importing flask.ext.migrate is deprecated, use flask_migrate instead.
  from flask.ext.migrate import Migrate, MigrateCommand
/home/ubuntu/workspace/app/__init__.py:10: ExtDeprecationWarning: Importing flask.ext.mail is deprecated, use flask_mail instead.
  from flask.ext.mail import Mail, Message
Traceback (most recent call last):
  File "manage.py", line 4, in <module>
    from app.models import User, Role
  File "/home/ubuntu/workspace/app/models.py", line 5, in <module>
    class Role(db.Model):
NameError: name 'db' is not defined
```
2. And here's the stack trace with `from . import db` included:
```
(venv)alwayslearnedstuff:~/workspace $ python manage.py shell
/home/ubuntu/workspace/app/__init__.py:3: ExtDeprecationWarning: Importing flask.ext.script is deprecated, use flask_script instead.
  from flask.ext.script import Manager
/home/ubuntu/workspace/app/__init__.py:5: ExtDeprecationWarning: Importing flask.ext.bootstrap is deprecated, use flask_bootstrap instead.
  from flask.ext.bootstrap import Bootstrap
/home/ubuntu/workspace/app/__init__.py:6: ExtDeprecationWarning: Importing flask.ext.moment is deprecated, use flask_moment instead.
  from flask.ext.moment import Moment
/home/ubuntu/workspace/app/__init__.py:7: ExtDeprecationWarning: Importing flask.ext.wtf is deprecated, use flask_wtf instead.
  from flask.ext.wtf import Form
/home/ubuntu/workspace/app/__init__.py:8: ExtDeprecationWarning: Importing flask.ext.sqlalchemy is deprecated, use flask_sqlalchemy instead.
  from flask.ext.sqlalchemy import SQLAlchemy
/home/ubuntu/workspace/app/__init__.py:9: ExtDeprecationWarning: Importing flask.ext.migrate is deprecated, use flask_migrate instead.
  from flask.ext.migrate import Migrate, MigrateCommand
/home/ubuntu/workspace/app/__init__.py:10: ExtDeprecationWarning: Importing flask.ext.mail is deprecated, use flask_mail instead.
  from flask.ext.mail import Mail, Message
/home/ubuntu/workspace/venv/local/lib/python2.7/site-packages/flask_sqlalchemy/__init__.py:794: FSADeprecationWarning: SQLALCHEMY_TRACK_MODIFICATIONS adds significant overhead and will be disabled by default in the future.  Set it to True or False to suppress this warning.
  'SQLALCHEMY_TRACK_MODIFICATIONS adds significant overhead and '
Traceback (most recent call last):
  File "manage.py", line 11, in <module>
    app = create_app(os.getenv('FLASK_CONFIG') or 'default')
  File "/home/ubuntu/workspace/app/__init__.py", line 32, in create_app
    from .main import main as main_blueprint
ImportError: No module named main
```

**Comment by miguelgrinberg** [comment]
And you are positive you have a file named `/home/ubuntu/workspace/app/main/__init__.py`? The error is pretty clear, you have some sort of problem with the way you have structured your files.

**Comment by alwayslearnedstuff** [comment]
ok - thank you for looking ; I'll take another look at all the files.

**Comment by alwayslearnedstuff** [comment]
Update: fixed and got these two commands to work (which completes Ch7)! 
`python manage.py test`
`python manage.py db upgrade`

Here's what was wrong: 
1. I misnamed app/main/__init__.py     as app/main/__init.py__     oops! Took awhile to see that one. 

... but then I was getting a `Form` not recognized as a variable which led this issue:

2. `forms.py` is missing from Ch7. I went over Ch7 at least half dozen times to see if I had missed something--and even back to Ch4 to see if forms.py was there, similar to how models.py was introduced in Ch5--but it wasn't. No worries, no blame: all good and got me to think through it. I added this code to get it to work: 
```
from . import main
from flask.ext.wtf import Form
from wtforms import StringField, SubmitField
from wtforms.validators import Required

class NameForm(Form):
  name = StringField('What is your name?', validators=[Required()])
  submit = SubmitField('Submit')
```

Stack trace was helpful - thanks again! 

Resolved.
